qutip/solver/integrator.py

```diff
@@ -0,0 +1,185 @@
+""" `Integrator`: ODE solver wrapper to use in qutip's Solver """
+
+
+__all__ = ['Integrator', 'IntegratorException', 'add_integrator',
+           'sesolve_integrators', 'mesolve_integrators', 'mcsolve_integrators']
+
+
+from itertools import product
+from qutip import QobjEvo, qeye, basis
+from functools import partial
+
+
+class IntegratorException(Exception):
+    pass
+
+
+class Integrator:
+    """
+    A wrapper around ODE solvers.
+    It ensures a common interface for Solver usage.
+    It takes and return states as :class:`qutip.core.data.Data`, it may return
+    a different data-type than the input type.
+
+    Parameters
+    ----------
+    sytem: qutip.QobjEvo
+        Quantum system in which states evolve.
+
+    options: qutip.SolverOptions
+        Options for the solver.
+    """
+    # Used options in qutip.SolverOdeOptions
+    used_options = []
+    # Can evolve time dependent system
+    support_time_dependant = None
+    # Use the QobjEvo's matmul_data method as the driving function
+    use_QobjEvo_matmul = None
+
+    def __init__(self, system, options):
+        self.system = system
+        self._stats = {}
```

Comment:
I wanted to have stats, like total number of steps, average step length, number of failed steps, etc. It would help debug and optimize integration.
But it's not that simple to extract them from scipy, so I removed it at the last minute, but I still want it...
I can have them for solver we make ourselves, so it could stay empty for scipy integrator and but be in for verner methods... Or we drop it altogether...
I am removing it for now and think about it later, it's really just a nice to have but not worth delaying v5 for.

Reply:
Having stats seems like a really good idea. Maybe let's leave it for another PR though because it requires a bit of thinking. It would also help with debugging (e.g. we could ask people file GitHub issues to post their solver stats).